### PR TITLE
Add ROKS satellite cluster resource

### PR DESCRIPTION
**Pull request by anilkumarnagaraj**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

File: ibm/resource_ibm_satellite_cluster.go [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The unique name for the new IBM Cloud Satellite cluster",
+			},
+			"location": {
+				Type:        schema.TypeString,
+				Required:    true,
```
Comment (kavya498) [inline ibm/resource_ibm_satellite_cluster.go:54]: I believe location should be a force New resource

**Review by pierre-archambault — COMMENTED** [review]
Please see my comments

File: ibm/resource_ibm_satellite_cluster.go [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+				"Error updating the labels: %s\n%s", err, response)
+		}
+	}
+
+	//Wait for cluster to get warning state
+	_, err = waitForClusterToReady(*instance.ID, d, meta)
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for getting cluster (%s) to be warning state: %s", *instance.ID, err)
+	}
```
Comment (hkantare) [inline ibm/resource_ibm_satellite_cluster.go:354]: Need to add code to attach tags

File: ibm/resource_ibm_satellite_cluster.go [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+
+		cluster, response, err := satClient.GetCluster(getSatClusterOptions)
+		if err != nil {
+			return fmt.Errorf("Error retrieving cluster %s: %s", clusterID, err)
+		}
+		waitForWorkerUpdate := d.Get("wait_for_worker_update").(bool)
+
+		for _, w := range workerFields {
+			//kubeversion update done if there is a change in Major.Minor version
+			if strings.Split(*w.KubeVersion.Actual, "_")[0] != strings.Split(*cluster.MasterKubeVersion, "_")[0] {
```
Comment (hkantare) [inline ibm/resource_ibm_satellite_cluster.go:487]: I think we need to add more additional condition similar to container cluster

File: ibm/resource_ibm_satellite_cluster.go [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+					Action: "update",
+				}
+				err = wrkAPI.Update(clusterID, *w.ID, params, targetEnv)
+				if err != nil {
+					d.Set("patch_version", nil)
+					return fmt.Errorf("Error updating worker %s: %s", *w.ID, err)
+				}
+				if waitForWorkerUpdate {
+					_, err = WaitForSatelliteWorkerAvailable(d, meta, targetEnv)
+					if err != nil {
```
Comment (hkantare) [inline ibm/resource_ibm_satellite_cluster.go:498]: here also add 
d.Set("patch_version", nil)

File: ibm/resource_ibm_satellite_cluster.go [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+			Labels:             labels,
+			XAuthResourceGroup: &resourceGroup,
+		}
+		response, err := satClient.V2SetWorkerPoolLabels(wpots)
+		if err != nil {
+			return fmt.Errorf(
+				"Error updating the labels: %s\n%s", err, response)
+		}
+	}
+
```
Comment (hkantare) [inline ibm/resource_ibm_satellite_cluster.go:543]: Add the update support to add or remove additional zones to the cluster

File: ibm/data_source_ibm_satellite_cluster.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+							Type:     schema.TypeList,
+							Computed: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"zone": {
+										Type:     schema.TypeString,
+										Computed: true,
+									},
+									"worker_count": {
+										Type:     schema.TypeInt,
```
Comment (hkantare) [inline ibm/data_source_ibm_satellite_cluster.go:90]: worker_count to workers (follow same naming convention across all)

File: ibm/resource_ibm_satellite_cluster.go [hunk 7]
```diff
@@ -0,0 +1,766 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"enable_config_admin": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Optional:    true,
+				Description: "Grant cluster admin access to Satellite Config to manage Kubernetes resources.",
+			},
+			"workers": {
```
Comment (hkantare) [inline ibm/resource_ibm_satellite_cluster.go:117]: change to worker_count to have similar naming convention across all resources/datasources

File: ibm/resource_ibm_satellite_location.go [hunk 8]
```diff
@@ -123,6 +123,13 @@ func resourceIBMSatelliteLocation() *schema.Resource {
 				Computed:    true,
 				Description: "ID of the resource group.",
 			},
+			"tags": {
+				Type:     schema.TypeSet,
```
Comment (hkantare) [inline ibm/resource_ibm_satellite_location.go:127]: remove this changes its handled as part of other pr

File: ibm/resource_ibm_satellite_cluster.go [hunk 9]
```diff
@@ -0,0 +1,766 @@
...
+
+		response, err := satClient.V2SetWorkerPoolLabels(wpots)
+		if err != nil {
+			return fmt.Errorf(
+				"Error updating the labels: %s\n%s", err, response)
+		}
+	}
+
+	//Wait for cluster to get warning state
+	_, err = waitForClusterToReady(*instance.ID, d, meta)
```
Comment (hkantare) [inline ibm/resource_ibm_satellite_cluster.go:350]: Can we have wait_till parameter and wait for either one worker node ready or ingress ready similar to other clusters

File: ibm/resource_ibm_satellite_cluster.go (hunk 1 above)
Comment (anilkumarnagaraj) [reply ibm/resource_ibm_satellite_cluster.go:65]: location attribute is immutable now.

File: examples/ibm-satellite/modules/cluster/README.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+  cluster           = var.cluster
+  location          = var.location
+  kube_version      = var.kube_version
+  zones             = var.location_zones
+  resource_group    = var.resource_group
+}
+```
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+## Inputs
```
[the remark this answers is not in the document]
Comment (anilkumarnagaraj) [reply examples/ibm-satellite/modules/cluster/README.md:46]: Added attribute reference for cluster resource

File: examples/ibm-satellite/modules/cluster/variables.tf [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  type         = string
+}
+
+variable "kube_version" {
+  description = "Satellite Kube Version"
+  type         = string
+}
+
+variable "resource_group" {
+  description = "Resource Group Name that has to be targeted"
```
[the remark this answers is not in the document]
Comment (anilkumarnagaraj) [reply examples/ibm-satellite/modules/cluster/variables.tf:17]: Added resource_group_id attribute to cluster resource

File: ibm/data_source_ibm_satellite_cluster.go [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+)
+
+func dataSourceIBMSatelliteCluster() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceIBMSatelliteClusterRead,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "Name or id of the cluster",
+				Type:        schema.TypeString,
```
[the remark this answers is not in the document]
Comment (anilkumarnagaraj) [reply ibm/data_source_ibm_satellite_cluster.go:22]: 'name' is required parameter now.

File: ibm/data_source_ibm_satellite_cluster.go [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"region": {
+				Type:        schema.TypeString,
+				Optional:    true,
```
[the remark this answers is not in the document]
Comment (anilkumarnagaraj) [reply ibm/data_source_ibm_satellite_cluster.go]: Remoed region

File: ibm/resource_ibm_satellite_cluster.go (hunk 2 above)
Comment (anilkumarnagaraj) [reply ibm/resource_ibm_satellite_cluster.go:395]: Attached tags to cluster

File: ibm/resource_ibm_satellite_cluster.go (hunk 3 above)
Comment (anilkumarnagaraj) [reply ibm/resource_ibm_satellite_cluster.go]: Added additional parameter to validate

File: ibm/resource_ibm_satellite_cluster.go (hunk 4 above)
Comment (anilkumarnagaraj) [reply ibm/resource_ibm_satellite_cluster.go:560]: Added d.Set("patch_version", nil)

File: ibm/resource_ibm_satellite_cluster.go (hunk 5 above)
Comment (anilkumarnagaraj) [reply ibm/resource_ibm_satellite_cluster.go:606]: Update method will support zones update.

File: ibm/resource_ibm_satellite_cluster.go (hunk 7 above)
Comment (anilkumarnagaraj) [reply ibm/resource_ibm_satellite_cluster.go]: Changed to worker_count

File: ibm/resource_ibm_satellite_location.go (hunk 8 above)
Comment (anilkumarnagaraj) [reply ibm/resource_ibm_satellite_location.go]: Removed tags

File: ibm/resource_ibm_satellite_cluster.go (hunk 9 above)
Comment (anilkumarnagaraj) [reply ibm/resource_ibm_satellite_cluster.go:417]: Fixed

**Pull request merged**